src/nvim/search.c

```diff
@@ -5106,3 +5106,527 @@ bool search_was_last_used(void)
+      l = fmatchlist;
+    }
+
+    // Copy the matching strings with a valid score to the return list
+    for (i = 0; i < len; i++) {
+      if (ptrs[i].score == SCORE_NONE) {
+        break;
+      }
+      tv_list_append_tv(l, TV_LIST_ITEM_TV(ptrs[i].item));
+    }
+
+    // next copy the list of matching positions
+    if (retmatchpos) {
+      li = tv_list_find(fmatchlist, -1);
+      tv_li = TV_LIST_ITEM_TV(li);
+      if (li == NULL || tv_li->vval.v_list == NULL) {
+        goto done;
+      }
+      l = tv_li->vval.v_list;
+
+      for (i = 0; i < len; i++) {
+        if (ptrs[i].score == SCORE_NONE) {
+          break;
+        }
+        if (ptrs[i].lmatchpos != NULL) {
+          tv_list_append_list(l, ptrs[i].lmatchpos);
+        }
+      }
+    }
+  }
+
+done:
+  xfree(ptrs);
+}
+
+// Do fuzzy matching. Returns the list of matched strings in 'rettv'.
+// If 'retmatchpos' is TRUE, also returns the matching character positions.
+void do_fuzzymatch(typval_T *argvars, typval_T *rettv, int retmatchpos)
+{
+  Callback cb = CALLBACK_NONE;
+  char *key = NULL;
+
+  // validate and get the arguments
+  if (argvars[0].v_type != VAR_LIST || argvars[0].vval.v_list == NULL) {
+    EMSG2(_(e_listreq), retmatchpos ? "matchfuzzypos()" : "matchfuzzy()");
+    return;
+  }
+  if (argvars[1].v_type != VAR_STRING
+      || argvars[1].vval.v_string == NULL) {
+    EMSG2(_(e_invarg2), tv_get_string(&argvars[1]));
+    return;
+  }
+
+  if (argvars[2].v_type != VAR_UNKNOWN) {
+    dict_T *d;
+    dictitem_T *di;
+
+    if (argvars[2].v_type != VAR_DICT || argvars[2].vval.v_dict == NULL) {
+      EMSG(_(e_dictreq));
+      return;
+    }
+
+    // To search a dict, either a callback function or a key can be
+    // specified.
+    d = argvars[2].vval.v_dict;
+    if ((di = tv_dict_find(d, "key", -1)) != NULL) {
+      if (di->di_tv.v_type != VAR_STRING
+          || di->di_tv.vval.v_string == NULL
+          || *di->di_tv.vval.v_string == NUL) {
+        EMSG2(_(e_invarg2), tv_get_string(&di->di_tv));
+        return;
+      }
+      key = (char *)tv_get_string(&di->di_tv);
+    } else if ((di = tv_dict_find(d, "text_cb", -1)) != NULL) {
+      if (callback_from_typval(&cb, &di->di_tv)) {
+        EMSG2(_(e_invargval), "text_cb");
+        return;
+      }
+    }
+  }
+
+  // get the fuzzy matches
+  tv_list_alloc_ret(rettv, kListLenMayKnow);
+
+  if (retmatchpos) {
+    list_T *l;
+
+    // For matchfuzzypos(), a list with two items are returned. First item
+    // is a list of matching strings and the second item is a list of
+    // lists with matching positions within each string.
+    l = tv_list_alloc(kListLenShouldKnow);
+    if (l == NULL) {
+      goto done;
+    }
+
+    tv_list_append_list(rettv->vval.v_list, l);
+
+    l = tv_list_alloc(kListLenShouldKnow);
+    if (l == NULL) {
+      goto done;
+    }
+
+    tv_list_append_list(rettv->vval.v_list, l);
+  }
+
+  char buf[NUMBUFLEN];
+  if (tv_get_string_buf_chk(&argvars[1], buf)) {
```

Comment:
Shouldn't you be using the return value of `tv_get_string_buf_chk()`? It uses `buf` *only* if `&argvars[1]` is not a string, IIUC.

Reply:
I don't really know tbh, I actually never ported patches from upstream vim

Reply:
Source says that you should must use the return value:

https://github.com/neovim/neovim/blob/4e6f00dd29332ce549006e8df1b1392ed4209954/src/nvim/eval/typval.c#L2967-L2980